rfd/0192-oci-join.md

```diff
@@ -0,0 +1,205 @@
+        # ...
+```
+
+#### Join process
+
+When a node initiates the Oracle join method:
+
+- The node first fetches credentials for its
+[instance principal](https://docs.oracle.com/en-us/iaas/Content/Identity/Tasks/callingservicesfrominstances.htm#concepts)
+via the Oracle instance metadata service.
+- With those credentials, the node will create a
+[signed HTTP request](https://docs.oracle.com/en-us/iaas/Content/API/Concepts/signingrequests.htm)
+to the Oracle Cloud API to fetch the compartment the instance is in, at
+`https://iaas.{region}.oraclecloud.com/{api version}/compartment/{compartment's OCID}`.
+The instance's principal does not need any additional permissions to make this request.
+- The node will then make a `RegisterUsingToken` request to the auth server and
+sends the URL and headers of the signed request as parameters (the auth server
+will be able to recreate the rest of the request).
+
+When the auth server receives a `RegisterUsingToken` request for the Oracle join method:
+
+- The auth server extracts the key ID from the provided Authorization header. The key
+ID is a string that Oracle uses to identify the caller. For instance principals,
+the key ID is a JWT prefixed by the string `ST$` (unfortunately I could not
+find docs that back this up, I found this experimentally).
+- The auth server decodes the JWT and maps the claims `opc-tenant`,
+`opc-compartment`, and `opc-instance` to the instance's tenancy ID, compartment
+ID, and instance ID respectively.
+- The auth server verifies that the compartment ID and region in the provided
+API URL match the compartment ID and region from the claims (the region can be
+extracted from the instance ID).
+- The auth server reconstructs and performs the API request; if Oracle accepts
+it, the auth server validates the tenancy ID, compartment ID (or name, found in
+the API response), and region against the Teleport provision token. If that
+passes, the node is allowed to join the cluster.
+
+#### Limitations
+
+The Oracle provision tokens will not support nested compartments, i.e. if
+compartment `foo` has a child compartment `bar` and the provision token has
+`compartments: ["foo"]`, this will not allow instances in container `bar` to
+join. This is for simplicity's sake; Teleport would need to make several
+requests to the Oracle Cloud API to walk up the compartment tree from the
+compartment the instance is in, each of which would need to be signed. This
+would require a complicated back-and-forth between the auth server and the
+joining node to get signed requests for each compartment.
+
+### Security
+
+Teleport will not at any point verify the claims in the key ID JWT provided by
+the instance. This is because the needed JWKs will always be behind a
+[non-public API](https://docs.oracle.com/en-us/iaas/Content/APIGateway/Tasks/apigatewayusingjwttokens.htm#identityprovider).
+Instead, Teleport will trust the response from the Oracle Cloud API to know if
+it can trust the claims.
```

Comment:
Anything to consider re: SSRF?

If we're building the API URL from user-provided region and OCID values, can we sanitize those values first?

Reply:
👍  Would be good to see this acknowledged as a security concern/consideration within the RFD.

Reply:
I added a note in the security section about sanitizing values, let me know if it needs more details.